torch/distributed/_shard/checkpoint/state_dict_saver.py

```diff
@@ -0,0 +1,150 @@
+import io
+from typing import Any, Dict, List, Tuple
+
+import torch
+import torch.distributed as dist
+from torch import Tensor
+from torch.distributed._shard.sharded_tensor import (
+    ShardedTensor,
+)
+
+from .metadata import (
+    Metadata,
+    BytesWriteRequest,
+    TensorStorageMetadata,
+    TensorWriteRequest,
+)
+from .resharding import _prepare_sharded_tensor_write
+from .storage import StorageWriter
+
+# -------------- private functions --------------
+def _compute_tensor_md(fqn: str, tensor: Tensor) -> TensorStorageMetadata:
+    return TensorStorageMetadata(
+        storage_key=fqn,
+        size=tensor.size()
+    )
+
+
+def _prepare(
+    state_dict: Dict[str, Any], include_non_replicated_tensors: bool = False
+) -> Tuple[Metadata, Dict[str, int], List[BytesWriteRequest], List[TensorWriteRequest]]:
+    """
+    Build a serialization plan for a given state_dict
+    By default, regular tensors are only included if execution from `rank 0`.
+
+    Args:
+        state_dict: The instance to plan for.
+        include_non_replicated_tensors: Include non-sharded tensors even if rank != 0
```

Comment:
nit: We are calling it non-sharded in one place and non-replicated in another place. Lets have consistent terminology.

Reply:
I replaced non-sharded to be more specific to non ShardedTensor items.